lib/spack/spack/repository.py

```diff
@@ -112,48 +128,138 @@ def __init__(self, *repo_dirs, **kwargs):
         self._all_package_names = None
         self._provider_index = None
 
-        # If repo_dirs is empty, just use the configuration
-        if not repo_dirs:
-            import spack.config
-            repo_dirs = spack.config.get_config('repos')
-            if not repo_dirs:
-                raise NoRepoConfiguredError(
-                    "Spack configuration contains no package repositories.")
-
         # Add each repo to this path.
         for root in repo_dirs:
-            try:
-                repo = Repo(root, self.super_namespace)
-                self.put_last(repo)
-            except RepoError as e:
-                tty.warn("Failed to initialize repository at '%s'." % root,
-                         e.message,
-                         "To remove the bad repository, run this command:",
-                         "    spack repo rm %s" % root)
+            self.append_from_path(root)
+
+    def append_from_path(self, root):
+        """Append a repository to the list of managed ones. Takes a path
+        as argument.
+
+        Args:
+            root: path to the repository
+        """
+        try:
+            repo = Repo(root, self.super_namespace)
+            self.append(repo)
```

Comment:
With an IDE this is a <kbd>Ctrl</kbd>+<kbd>B</kbd> away:
```python
def append(self, value):
    'S.append(value) -- append value to the end of the sequence'
    self.insert(len(self), value)
```
> However while it shouldn't make this class harder to use, IMO it does make it harder to modify.

It does make the class *easier to use*, as it is now adhering to a stronger contract - that of `MutableSequence`s. I don't get you point when you say it will be harder to modify. Why?